internal/dataplane/parser/translators/kongupstreampolicy.go

```diff
@@ -0,0 +1,147 @@
+package translators
+
+import (
+	"github.com/kong/go-kong/kong"
+	"github.com/samber/lo"
+
+	kongv1beta1 "github.com/kong/kubernetes-ingress-controller/v2/pkg/apis/configuration/v1beta1"
+)
+
+// TranslateKongUpstreamPolicy translates KongUpstreamPolicySpec to kong.Upstream. It makes assumption that
+// KongUpstreamPolicySpec has been validated on the API level.
+func TranslateKongUpstreamPolicy(policy kongv1beta1.KongUpstreamPolicySpec) *kong.Upstream {
+	return &kong.Upstream{
+		Algorithm:    policy.Algorithm,
+		Slots:        policy.Slots,
+		Healthchecks: translateHealthchecks(policy.Healthchecks),
+		HostHeader:   policy.HostHeader,
+
+		HashOn:           translateHashOn(policy.HashOn),
+		HashOnHeader:     translateHashOnHeader(policy.HashOn),
+		HashOnURICapture: translateHashOnURICapture(policy.HashOn),
+		HashOnCookie:     translateHashOnCookie(policy.HashOn),
+		HashOnCookiePath: translateHashOnCookiePath(policy.HashOn),
+		HashOnQueryArg:   translateHashOnQueryArg(policy.HashOn),
+
+		HashFallback:           translateHashOn(policy.HashOnFallback),
+		HashFallbackHeader:     translateHashOnHeader(policy.HashOnFallback),
+		HashFallbackURICapture: translateHashOnURICapture(policy.HashOnFallback),
+		HashFallbackQueryArg:   translateHashOnQueryArg(policy.HashOnFallback),
+	}
+}
+
+func translateHashOn(hashOn *kongv1beta1.KongUpstreamHash) *string {
+	if hashOn == nil {
+		return nil
+	}
+	// Only one of hashOn fields can be set.
+	switch {
+	case hashOn.Header != nil:
```

Comment:
Why this order? This specifies a priority on these fields, so it will be clearer if we add some comments to tell where does the priority come from.

Reply:
The order here will be irrelevant when https://github.com/Kong/kubernetes-ingress-controller/issues/4951 is implemented - that's why the comment `Only one of hashOn fields can be set`. I changed that to be more elaborative.